### PR TITLE
Document Lens functions.

### PR DESCRIPTION
**Pull request by vic**
Added documentation with a simple example on each so that users
can get a quick glimpse at how each lens works.

As we are still on `v0` I made some changes to make the functions
be more Elixir way, for example, made `to_list`, `map`, `each`, and
`get_and_update` take the lens as first parameter, to ease piping
and to be more consistent with mapping collections in elixir.

```elixir
Lens.filter(&Integer.is_odd/1) |> Lens.map([1, 2], fn v -> v + 10 end)
```

Also added `Lens.get` which is just `Lens.list` but unwraps the
value if the list length is one, otherwise returns the whole list.

### CODE REVIEW

**Comment by vic** [comment]
Hi @obrok , have just seen your latest commit @c63fea83c424cc3437b760061550a2c32154670d can merge that if you want, I made lots of changes here, could you review them and tell me what you think?

The `master` branch on my fork has all my PR applied and conflicts resolved.

**Comment by vic** [comment]
If you feel like merging this we can work on improving my documentation wording, as yours is much much better than mine :)

**Comment by obrok** [comment]
This is cool, although I'm not really on board with the change in argument order. My reasoning is that most often you will define your lens somewhere and then actually want to pipe your datastructure _through_ a lens operation. Something like:

``` elixir
params
|> Database.get_some_data
|> Lens.to_list(MyLenses.the_part_of_data_im_interested_in)
|> some_further_processing
```

What do you think?

For the future: it would be much easier if you made separate commits or even PRs from conceptually independent things like adding a bunch of documentation and changing or extending the API.

**Comment by vic** [comment]
Yeah sorry about this PR that started out as documentation and later implemented more things on it. This is the last one:

I made the lens function compatible with `Kernel.get_and_update_in`. Now you can give any lens
directly as a key to `Kernel.get_and_update_in` and its friends `Kernel.get_in`, `Kernel.update_in`.
So your pipe example could look like the following by using these kernel functions:

``` elixir
params
|> Database.get_some_data
|> get_in([ MyLenses.the_part_of_data_im_interested_in ])
|> some_further_processing
```

Also implemented `to_list`, `each`, `map` in terms of these kernel functions. 
So basically `to_list` is `Kernel.get_in`, `map` is `Kernel.update_in`

File: lib/lens.ex [hunk 1]
```diff
@@ -30,12 +30,30 @@ defmodule Lens do
     end
   end
 
+  @doc ~S"""
```
Comment (obrok) [inline lib/lens.ex:33]: Please update all docstrings to match this style

File: lib/lens/macros.ex [hunk 2]
```diff
@@ -13,7 +13,16 @@ defmodule Lens.Macros do
     end
 
     quote do
-      def unquote(header), do: unquote(body)
+      def unquote(header) do
+        lens = unquote(body)
+        fn
+          :get, data, next ->
```
Comment (obrok) [inline lib/lens/macros.ex:19]: This is pretty neat. However, it has the side effect of making it worse suited for a public API. I envisioned clients using `deflens` in their own projects to get `|>` for free, but now that's not as simple (because of `wrap`). It seems there should be 2 macros - one to build this `:get/:get_and_update` fork internally and another one to expose to the world and use in cases like

``` elixir
  deflens all, do: wrap filter(fn _ -> true end)
```

so that `wrap` is not needed.

File: lib/lens.ex [hunk 3]
```diff
@@ -128,26 +215,59 @@ defmodule Lens do
...
+      iex> Lens.filter(&Integer.is_odd/1) |> Lens.map(data, fn v -> v + 10 end)
+      [11, 2, 13, 4]
+  """
+  @spec map(t, any, (any -> any)) :: any
+  def map(lens, data, fun), do: update_in(data, [lens], fun)
+
+  @doc ~S"""
+  Get a tuple of original values and the updated data by applying fun on lens.
+
+  The mapping function takes a value and must return a tuple with old and update values.
```
Comment (obrok) [inline lib/lens.ex]: The first item in the tuple is the value to return, not the old value.

File: lib/lens.ex [hunk 4]
```diff
@@ -128,26 +215,59 @@ defmodule Lens do
...
+
+      iex> data = [1, 2, 3]
+      iex> Lens.filter(&Integer.is_odd/1) |> Lens.get_and_map(data, fn v -> {v, v + 10} end)
```
Comment (obrok) [inline lib/lens.ex]:
``` elixir
iex> Lens.filter(&Integer.is_odd/1) |> Lens.get_and_map(data, fn v -> {v + 1, v + 10} end)
{[2, 4], [11, 2, 13]}
```

would maybe be a better example

**Review by obrok — CHANGES_REQUESTED** [review]

File: lib/lens/macros.ex (hunk 2 above)
Comment (obrok) [reply lib/lens/macros.ex:19]: Ok, maybe let's finish up this PR first, because it's already a monster, and then I can figure out what to do with these macros. Please address the other comments, I made and then I can merge. I think I came around about the order of arguments to `Lens.map`, etc. `Regex` also has this order even though it's really weird to use, so I guess we're in good company.

File: lib/lens.ex (hunk 4 above)
Comment (vic) [reply lib/lens.ex]: done, improved the example

File: lib/lens/macros.ex (hunk 2 above)
Comment (vic) [reply lib/lens/macros.ex:19]: You are right about `wrap`, so let's just finish this PR (just fixed your comments) and we can work on another issue.

**Review by obrok — APPROVED** [review]

File: lib/lens.ex [hunk 5]
```diff
@@ -3,7 +3,7 @@ defmodule Lens do
 
   @opaque t :: (:get, any, function -> list(any)) | (:get_and_update, any, function -> {list(any), any})
 
-  @doc """
+  @doc ~S"""
```
Comment (obrok) [inline lib/lens.ex:6]: Should have made 2 commits here. A commit with "and" in the title is a red flag.
